pkg/controllers/pdgroup/tasks/status.go

```diff
@@ -15,82 +15,129 @@
+		suspendMessage := "pd group is suspending"
 
-func NewTaskStatus(logger logr.Logger, c client.Client) task.Task[ReconcileContext] {
-	return &TaskStatus{
-		Client: c,
-		Logger: logger,
-	}
-}
+		suspended := true
+		pdg := state.PDGroup()
+		pds := state.PDSlice()
+		for _, pd := range pds {
+			if !meta.IsStatusConditionTrue(pd.Status.Conditions, v1alpha1.PDCondSuspended) {
+				suspended = false
+			}
+		}
+		if suspended {
+			suspendStatus = metav1.ConditionTrue
+			suspendMessage = "pd group is suspended"
+		}
+
+		needUpdate := meta.SetStatusCondition(&pdg.Status.Conditions, metav1.Condition{
+			Type:               v1alpha1.PDGroupCondSuspended,
+			Status:             suspendStatus,
+			ObservedGeneration: pdg.Generation,
+			Reason:             v1alpha1.PDGroupSuspendReason,
+			Message:            suspendMessage,
+		})
+		needUpdate = SetIfChanged(&pdg.Status.ObservedGeneration, pdg.Generation) || needUpdate
 
-func (*TaskStatus) Name() string {
-	return "Status"
+		if needUpdate {
+			if err := c.Status().Update(ctx, state.PDGroup()); err != nil {
+				return task.Fail().With("cannot update status: %v", err)
+			}
+		}
+
+		return task.Complete().With("status of suspend pd is updated")
+	})
 }
 
-func (t *TaskStatus) Sync(ctx task.Context[ReconcileContext]) task.Result {
-	rtx := ctx.Self()
+func TaskStatus(state *ReconcileContext, c client.Client) task.Task {
+	return task.NameTaskFunc("Status", func(ctx context.Context) task.Result {
+		pdg := state.PDGroup()
 
-	suspendStatus := metav1.ConditionFalse
-	suspendMessage := "pd group is not suspended"
-	if rtx.Suspended {
-		suspendStatus = metav1.ConditionTrue
-		suspendMessage = "pd group is suspended"
-	} else if rtx.Cluster.ShouldSuspendCompute() {
-		suspendMessage = "pd group is suspending"
-	}
-	conditionChanged := meta.SetStatusCondition(&rtx.PDGroup.Status.Conditions, metav1.Condition{
-		Type:               v1alpha1.PDGroupCondSuspended,
-		Status:             suspendStatus,
-		ObservedGeneration: rtx.PDGroup.Generation,
-		Reason:             v1alpha1.PDGroupSuspendReason,
-		Message:            suspendMessage,
+		suspendStatus := metav1.ConditionFalse
+		suspendMessage := "pd group is not suspended"
+		needUpdate := meta.SetStatusCondition(&pdg.Status.Conditions, metav1.Condition{
+			Type:               v1alpha1.PDGroupCondSuspended,
+			Status:             suspendStatus,
+			ObservedGeneration: pdg.Generation,
+			Reason:             v1alpha1.PDGroupSuspendReason,
+			Message:            suspendMessage,
+		})
+
+		replicas, readyReplicas, updateReplicas, currentReplicas := calcReplicas(state.PDSlice(), state.CurrentRevision, state.UpdateRevision)
+
+		// all instances are updated
+		if updateReplicas == replicas {
+			// update current revision
+			state.CurrentRevision = state.UpdateRevision
+			// update status of pdg version
+			// TODO(liubo02): version of a group is hard to understand
+			// We need to change it to a more meaningful field
```

Comment:
We need a field to indicate if a group is upgrading.

Reply:
prefer condition